File: src/org/cappuccino_project/ide/intellij/plugin/utils/ArrayUtils.java
package org.cappuccino_project.ide.intellij.plugin.utils;

import com.google.common.collect.ImmutableList;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

public class ArrayUtils {

    public static final List<String> EMPTY_STRING_ARRAY = ImmutableList.copyOf(new String[0]);

    /**
     * Joins
     * @param list list of strings to join
     * @return string of string elements joined by delimiter
     */
    public static String join(@NotNull final List<String> list) {
        return join(list, ", ");
    }


    /**
     * Joins
     * @param list list of strings to join
     * @param delimiter joining string
     * @return string of string elements joined by delimiter
     */
    public static String join(@NotNull final List<String> list, @NotNull final String delimiter) {
        return join(list, delimiter, false);
    }

    /**
     * Joins
     * @param list list of strings to join
     * @param delimiter joining string
     * @return string of string elements joined by delimiter
     */
    public static String join(@NotNull final List<String> list, @NotNull final String delimiter, boolean trailing) {
        final StringBuilder builder = new StringBuilder();
        for (final String string : list) {
            builder.append(string).append(delimiter);
        }
        String out = builder.toString();
        if (!trailing && out.length() > delimiter.length()) {
            out = out.substring(0, out.length()-delimiter.length());
        }
        return out;
    }

    /**
     * Filters array list and returns a subset of items matching a given class
     * @param list list of items
     * @param filterClass class to filter items for
     * @param <T> type of element to return
     * @return list of items matching class
     */
    public static <T> List<T> filter(List<?> list, Class<T> filterClass) {
        if (list == null || list.isEmpty()) {
            return Collections.emptyList();
        }
        List<T> out = new ArrayList<>();
        for (Object ob : list) {
            if (filterClass.isInstance(ob)) {
                out.add(filterClass.cast(ob));
            }
        }
        return out;
    }

    public static <T> List<T> filter(@NotNull List<T> list, @NotNull Filter<T> filter) {
        List<T> out = new ArrayList<>();
        for (T item : list) {
            if (filter.check(item)) {
                out.add(item);
            }
        }
        return out;
    }


    public static interface Filter<T> {
        boolean check(T item);
    }


    @NotNull
    public static List<String> search(@NotNull List<String> keywords, @NotNull String queryString) {
        List<String> out = new ArrayList<>();
        if (queryString.isEmpty()) {
            return Collections.emptyList();
        }
        int queryStringLength = queryString.length();
        for (String keyword : keywords) {
            if (keyword.length()>=queryStringLength && keyword.contains(queryString)) {
                out.add(keyword);
            }
        }
        return out;
    }

    @NotNull
    public static <K,T> List<T> flatten(Map<K,List<T>> map) {
        List<T> out = new ArrayList<>();
        for (K key : map.keySet()) {
            out.addAll(map.get(key));
        }
        return out;
    }

}

File: src/cappuccino/ide/intellij/plugin/utils/ArrayUtils.kt
package cappuccino.ide.intellij.plugin.utils

import com.intellij.openapi.util.MultiValuesMap
import java.util.ArrayList

typealias Filter<T> = (T) -> Boolean
object ArrayUtils {

    val EMPTY_STRING_ARRAY: List<String> = emptyList()

    /**
     * Joins
     * @param list list of strings to join
     * @param delimiter joining string
     * @return string of string elements joined by delimiter
     */
    @JvmOverloads
    fun join(list: List<String>, delimiter: String = ", ", trailing: Boolean = false): String {
        val builder = StringBuilder()
        for (string in list) {
            builder.append(string).append(delimiter)
        }
        var out = builder.toString()
        if (!trailing && out.length > delimiter.length) {
            out = out.substring(0, out.length - delimiter.length)
        }
        return out
    }

    /**
     * Filters array list and returns a subset of items matching a given class
     * @param list list of items
     * @param filterClass class to filter items for
     * @param <T> type of element to return
     * @return list of items matching class
    </T> */
    fun <T> filter(list: List<*>?, filterClass: Class<T>): List<T> {
        if (list == null || list.isEmpty()) {
            return emptyList()
        }
        val out = ArrayList<T>()
        for (ob in list) {
            if (filterClass.isInstance(ob)) {
                out.add(filterClass.cast(ob))
            }
        }
        return out
    }

    fun <T> filter(list: List<T>, filter: Filter<T>): List<T> {
        val out = ArrayList<T>()
        for (item in list) {
            if (filter(item)) {
                out.add(item)
            }
        }
        return out
    }




    fun search(keywords: List<String>, queryString: String): List<String> {
        val out = ArrayList<String>()
        if (queryString.isEmpty()) {
            return mutableListOf()
        }
        val queryStringLength = queryString.length
        for (keyword in keywords) {
            if (keyword.length >= queryStringLength && keyword.contains(queryString)) {
                out.add(keyword)
            }
        }
        return out
    }
}

fun List<String>.startsWith(prefix:String) : Boolean {
    this.forEach {
        if (it.startsWith(prefix))
            return true
    }
    return false;
}

fun List<String>.startsWithAny(prefixes:List<String>) : Boolean {
    prefixes.forEach {
        if (this.startsWith(it))
            return true
    }
    return false
}


fun <T> getFirstMatchOrNull(variableNameElements: List<T>, filter: Filter<T>): T? {
    return variableNameElements.firstOrNull(filter)
}
/*
fun <T> List<T>.plus(element:T) : List<T> {
    val list = toMutableList()
    list.add(element)
    return list
}

fun <T> List<T>.minus(element:T) : List<T> {
    if (!contains(element))
        return this
    val list = toMutableList()
    list.remove(element)
    return list
}*/

fun <Key, Type> MutableMap<Key, MutableList<Type>>.put(key:Key, item:Type) {
    if (!this.containsKey(key)) {
        this[key] = mutableListOf()
    }
    this[key]!!.add(item)
}

fun <T> Collection<T>?.isNotNullOrEmpty() : Boolean
    = this != null && this.isNotEmpty()
